mewarpx/mewarpx/emission.py

```diff
@@ -1277,6 +1414,148 @@ def get_normals(self, x, y, z):
+
+        self.area = 2.0 * np.pi * self.r * (self.zmax - self.zmin)
+
+        # Compute cell count
+        if mwxrun.geom_str == 'RZ':
+            self.cell_count = (self.zmax - self.zmin) / mwxrun.dz
+        elif mwxrun.geom_str == 'XYZ':
+            self.cell_count = (
+                self.area
+                / min(mwxrun.dx * mwxrun.dy, mwxrun.dx * mwxrun.dz,
+                      mwxrun.dy * mwxrun.dz)
+            )
+
+    def _get_xv_coords(self, npart, m, rseed):
+        """Get particle coordinates given particle number.
+
+        See :func:`mewarpx.emission.BaseEmitter.get_newparticles` for details.
+        """
+        if rseed is not None:
+            nprstate = np.random.get_state()
+            np.random.seed(rseed)
+            # rseedv is passed to get velocities. The basic rseed here is used
+            # for positions, below.
+            rseedv = np.random.randint(1000000000)
+        else:
+            rseedv = None
+
+        theta = np.random.uniform(0.0, 2.0*np.pi, npart)
+        cos_theta = np.cos(theta)
+        sin_theta = np.sin(theta)
+
+        x = self.r * cos_theta
+        y = self.r * sin_theta
+        z = np.random.uniform(self.zmin, self.zmax, npart)
+
+        vz, v_trans, v_long = mwxutil.get_velocities(
+            npart, self.T, m=m, transverse_fac=self.transverse_fac,
+            emission_type=self.emission_type, rseed=rseedv)
+
+        # Flip the longitudinal velocity if needed. This appears to be faster
+        # than an if statement for 10000 or fewer particles.
+        v_long = self.rsign * v_long
+
+        vx = cos_theta * v_long - sin_theta * v_trans
+        vy = sin_theta * v_long + cos_theta * v_trans
+
+        if rseed is not None:
+            np.random.set_state(nprstate)
+
+        return x, y, z, vx, vy, vz
+
+    def get_normals(self, x, y, z):
+        """Calculate local surface normal at specified coordinates.
+
+        Arguments:
+            x (np.ndarray): x-coordinates of emitted particles (in meters).
+            y (np.ndarray): y-coordinates of emitted particles (in meters).
+            z (np.ndarray): z-coordinates of emitted particles (in meters).
+
+        Returns:
+            normals (np.ndarray): nx3 array containing the outward surface
+                normal vector at each particle location.
+        """
+        normals = np.zeros((len(x), 3))
+
+        # Above the cylinder, the normal is just z-hat
+        normals[z > self.zmax, 2] = 1.0
+        # below the cylinder, the normal is just -z-hat
+        normals[z < self.zmin, 2] = -1.0
+
+        # On the side face, the normal is r-hat
+        normal_idxs = (z < self.zmax) & (z > self.zmin)
+        x_side = x[normal_idxs]
+        y_side = y[normal_idxs]
+        r_side = np.sqrt(x_side**2 + y_side**2)
+        normals[normal_idxs, 0] = x_side / r_side
+        normals[normal_idxs, 1] = y_side / r_side
+        return normals
```

Comment:
I would keep only the side face stuff here. That's because these are the normals for emitted electrons, and this class only emits from the side. Whereas you're porting from one that can actually emit from top & bottom faces.
```suggestion
        # The normal is r-hat
        r = np.sqrt(x**2 + y**2)
        normals[:, 0] = x / r
        normals[:, 1] = y / r
        return normals
```